common/lib/driver-definitions/src/storage.ts

```diff
@@ -118,36 +118,36 @@ export interface IDocumentStorageService extends Partial<IDisposable> {
     /**
      * Returns the snapshot tree.
      */
-    getSnapshotTree(version?: IVersion): Promise<ISnapshotTree | null>;
+    getSnapshotTree(version?: IVersion, fetchReason?: string): Promise<ISnapshotTree | null>;
```

Comment:
I'd not add it to all APIs. For example, uploadSummaryWithContext - not sure what kind of reason we can provide.
I'd start small - only with getSnapshotTree, getVersions, and maybe with readBlob - I'd leave the rest with no changes.
Even readBlob is questionable.

I'd also probably rename it and add comments to describe why and what it means (mention that's free-form string).
_callReason_ seems to be a better name. I'd look around for examples of something like that in other SDKs, if we can find (re naming).
BTW, are there any requirements for it? Like not having some special symbols?

Reply:
Sorry what do you mean by special symbols?

Reply:
is it ok to have symbols like = ,  ?
Looking at formatClientCorrelationId() I do not see any escaping logic, meaning using these symbols will screw up final string.

Reply:
Till now we didn't have a free form string. We either had boolean, uuid, number, version. I will add a comment here about the restriction of not having "=" and "," in the string and then would add some logic in the formatClientCorrelationId to handle such cases.

Reply:
Comments will be lost as we pass these arguments through layers. I think we should do better - find a way to escape such strings (or reject/drop them). Escaping is better, but assumes storage to understand how to un-escape.
I think we should either use something like encodeURIComponent() (and use same separators used with URIs, i.e. ? &) or simply drop "=" and "," signs from such strings when encoding (if we go that route, I'd stop adding spaces after "," and instruct SPO not to trim strings).

Reply:
I figured out we don't need to do this. X-RequestStats is not a custom header. It is used for these purposes generally by spo and it is recorded as a complete string instead of breaking down any separators. So, we can use any separator for beautifying it but otherwise it is not a problem. In that case I want to leave it as such as it is easier to read now as is.